src/S3/S3MultiRegionClient.php

```diff
@@ -224,17 +227,31 @@ private function determineRegionMiddleware() {
                         $region = null;
                         if (isset($result['@metadata']['headers']['x-amz-bucket-region'])) {
                             $region = $result['@metadata']['headers']['x-amz-bucket-region'];
+                            $this->cache->set($cacheKey, $region);
                         } else {
-                            /** @var S3ClientInterface $client */
-                            $client = $this->getClientFromPool();
-                            $region = (yield $client->determineBucketRegionAsync(
+                            $region = $this->getDetermineBucketRegionGenerator(
                                 $command['Bucket']
-                            ));
+                            );
                         }
 
-                        $this->cache->set($cacheKey, $region);
                         $command['@region'] = $region;
                         yield $handler($command);
+                    } catch (AwsException $e) {
+                        if ($e->getAwsErrorCode() === 'AuthorizationHeaderMalformed') {
+                            $region = $this->determineBucketRegionFromExceptionBody(
+                                $e->getResponse()->getBody()
+                            );
+                            if (!empty($region)) {
+                                $this->cache->set($cacheKey, $region);
+                            } else {
+                                $region = $this->getDetermineBucketRegionGenerator(
```

Comment:
Looks like calling `getDetermineBucketRegionGenerator` here is checking `AuthorizationHeaderMalformed` twice if cache is not found. This doesn't hurt, though might be worth considering separate the checking logic in my first comment?